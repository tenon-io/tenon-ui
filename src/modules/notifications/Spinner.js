import React, { Component } from 'react';
import uuidv4 from 'uuid/v4';
import PropTypes from 'prop-types';

/**
 * @component
 * An SVG spinner component. Props get spread onto the svg tag to allow
 * for adding css classes and aria
 *
 * @prop required {string} title - The alternative title for the spinner image.
 * @prop {string} className - The CSS class to apply to the svg element.
 *
 * @returns {React Element}
 */
class Spinner extends Component {
    static propTypes = {
        title: PropTypes.string,
        className: PropTypes.string
    };

    constructor(props) {
        super(props);

        this.svgTitleId = uuidv4();
    }

    render() {
        const { title, className, ...props } = this.props;
        return (
            <svg
                role="img"
                viewBox="0 0 58 58"
                xmlns="http://www.w3.org/2000/svg"
                aria-describedby={title ? this.svgTitleId : null}
                className={className ? className : null}
                {...props}
            >
                {title ? <title id={this.svgTitleId}>{title}</title> : null}
                <g fill="none" fillRule="evenodd">
                    <g
                        transform="translate(2 1)"
                        stroke="#FFFFFF"
                        strokeWidth="1.5"
                    >
                        <circle
                            cx="42.601"
                            cy="11.462"
                            r="5"
                            fillOpacity="1"
                            fill="#FFFFFF"
                        >
                            <animate
                                attributeName="fill-opacity"
                                begin="0s"
                                dur="1.3s"
                                values="1;0;0;0;0;0;0;0"
                                calcMode="linear"
                                repeatCount="indefinite"
                            />
                        </circle>
                        <circle
                            cx="49.063"
                            cy="27.063"
                            r="5"
                            fillOpacity="0"
                            fill="#FFFFFF"
                        >
                            <animate
                                attributeName="fill-opacity"
                                begin="0s"
                                dur="1.3s"
                                values="0;1;0;0;0;0;0;0"
                                calcMode="linear"
                                repeatCount="indefinite"
                            />
                        </circle>
                        <circle
                            cx="42.601"
                            cy="42.663"
                            r="5"
                            fillOpacity="0"
                            fill="#FFFFFF"
                        >
                            <animate
                                attributeName="fill-opacity"
                                begin="0s"
                                dur="1.3s"
                                values="0;0;1;0;0;0;0;0"
                                calcMode="linear"
                                repeatCount="indefinite"
                            />
                        </circle>
                        <circle
                            cx="27"
                            cy="49.125"
                            r="5"
                            fillOpacity="0"
                            fill="#FFFFFF"
                        >
                            <animate
                                attributeName="fill-opacity"
                                begin="0s"
                                dur="1.3s"
                                values="0;0;0;1;0;0;0;0"
                                calcMode="linear"
                                repeatCount="indefinite"
                            />
                        </circle>
                        <circle
                            cx="11.399"
                            cy="42.663"
                            r="5"
                            fillOpacity="0"
                            fill="#FFFFFF"
                        >
                            <animate
                                attributeName="fill-opacity"
                                begin="0s"
                                dur="1.3s"
                                values="0;0;0;0;1;0;0;0"
                                calcMode="linear"
                                repeatCount="indefinite"
                            />
                        </circle>
                        <circle
                            cx="4.938"
                            cy="27.063"
                            r="5"
                            fillOpacity="0"
                            fill="#FFFFFF"
                        >
                            <animate
                                attributeName="fill-opacity"
                                begin="0s"
                                dur="1.3s"
                                values="0;0;0;0;0;1;0;0"
                                calcMode="linear"
                                repeatCount="indefinite"
                            />
                        </circle>
                        <circle
                            cx="11.399"
                            cy="11.462"
                            r="5"
                            fillOpacity="0"
                            fill="#FFFFFF"
                        >
                            <animate
                                attributeName="fill-opacity"
                                begin="0s"
                                dur="1.3s"
                                values="0;0;0;0;0;0;1;0"
                                calcMode="linear"
                                repeatCount="indefinite"
                            />
                        </circle>
                        <circle
                            cx="27"
                            cy="5"
                            r="5"
                            fillOpacity="0"
                            fill="#FFFFFF"
                        >
                            <animate
                                attributeName="fill-opacity"
                                begin="0s"
                                dur="1.3s"
                                values="0;0;0;0;0;0;0;1"
                                calcMode="linear"
                                repeatCount="indefinite"
                            />
                        </circle>
                    </g>
                </g>
            </svg>
        );
    }
}

export default Spinner;
