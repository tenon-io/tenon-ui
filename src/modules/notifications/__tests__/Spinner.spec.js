jest.mock('uuid/v4');

import React from 'react';
import { render, cleanup } from 'react-testing-library';
import 'jest-dom/extend-expect';
import Spinner from '../Spinner';
import uuidv4 from 'uuid/v4';

describe('spinner', () => {
    afterEach(cleanup);

    it('should render a spinner', () => {
        uuidv4.mockReturnValueOnce('testId');

        const { container } = render(<Spinner title="Working" />);

        const svg = container.querySelector('svg');

        expect(svg).toHaveAttribute('aria-labelledby', 'testId');
        expect(svg).toHaveAttribute('role', 'img');

        const title = svg.querySelector('title');

        expect(title.innerHTML).toBe('Working');
        expect(title).toHaveAttribute('id', 'testId');

        expect(container).toMatchSnapshot();
    });

    it('should accept a custom CSS class', () => {
        uuidv4.mockReturnValueOnce('testId');

        const { container } = render(
            <Spinner title="Working" className="test-class" />
        );

        const svg = container.querySelector('svg');

        expect(svg).toHaveAttribute('class', 'test-class');
    });

    it('should also render a spinner without a title', () => {
        uuidv4.mockReturnValueOnce('testId');

        const { container } = render(<Spinner />);

        const svg = container.querySelector('svg');

        expect(svg).not.toHaveAttribute('aria-describedby');
        expect(svg).toHaveAttribute('role', 'img');

        const title = svg.querySelector('title');

        expect(title).toBeNull();

        expect(container).toMatchSnapshot();
    });
});
