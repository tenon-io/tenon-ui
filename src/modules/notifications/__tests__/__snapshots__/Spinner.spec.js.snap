// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`spinner should also render a spinner without a title 1`] = `
<div>
  <svg
    role="img"
    viewBox="0 0 58 58"
    xmlns="http://www.w3.org/2000/svg"
  >
    <g
      fill="none"
      fill-rule="evenodd"
    >
      <g
        stroke="#FFFFFF"
        stroke-width="1.5"
        transform="translate(2 1)"
      >
        <circle
          cx="42.601"
          cy="11.462"
          fill="#FFFFFF"
          fill-opacity="1"
          r="5"
        >
          <animate
            attributeName="fill-opacity"
            begin="0s"
            calcMode="linear"
            dur="1.3s"
            repeatCount="indefinite"
            values="1;0;0;0;0;0;0;0"
          />
        </circle>
        <circle
          cx="49.063"
          cy="27.063"
          fill="#FFFFFF"
          fill-opacity="0"
          r="5"
        >
          <animate
            attributeName="fill-opacity"
            begin="0s"
            calcMode="linear"
            dur="1.3s"
            repeatCount="indefinite"
            values="0;1;0;0;0;0;0;0"
          />
        </circle>
        <circle
          cx="42.601"
          cy="42.663"
          fill="#FFFFFF"
          fill-opacity="0"
          r="5"
        >
          <animate
            attributeName="fill-opacity"
            begin="0s"
            calcMode="linear"
            dur="1.3s"
            repeatCount="indefinite"
            values="0;0;1;0;0;0;0;0"
          />
        </circle>
        <circle
          cx="27"
          cy="49.125"
          fill="#FFFFFF"
          fill-opacity="0"
          r="5"
        >
          <animate
            attributeName="fill-opacity"
            begin="0s"
            calcMode="linear"
            dur="1.3s"
            repeatCount="indefinite"
            values="0;0;0;1;0;0;0;0"
          />
        </circle>
        <circle
          cx="11.399"
          cy="42.663"
          fill="#FFFFFF"
          fill-opacity="0"
          r="5"
        >
          <animate
            attributeName="fill-opacity"
            begin="0s"
            calcMode="linear"
            dur="1.3s"
            repeatCount="indefinite"
            values="0;0;0;0;1;0;0;0"
          />
        </circle>
        <circle
          cx="4.938"
          cy="27.063"
          fill="#FFFFFF"
          fill-opacity="0"
          r="5"
        >
          <animate
            attributeName="fill-opacity"
            begin="0s"
            calcMode="linear"
            dur="1.3s"
            repeatCount="indefinite"
            values="0;0;0;0;0;1;0;0"
          />
        </circle>
        <circle
          cx="11.399"
          cy="11.462"
          fill="#FFFFFF"
          fill-opacity="0"
          r="5"
        >
          <animate
            attributeName="fill-opacity"
            begin="0s"
            calcMode="linear"
            dur="1.3s"
            repeatCount="indefinite"
            values="0;0;0;0;0;0;1;0"
          />
        </circle>
        <circle
          cx="27"
          cy="5"
          fill="#FFFFFF"
          fill-opacity="0"
          r="5"
        >
          <animate
            attributeName="fill-opacity"
            begin="0s"
            calcMode="linear"
            dur="1.3s"
            repeatCount="indefinite"
            values="0;0;0;0;0;0;0;1"
          />
        </circle>
      </g>
    </g>
  </svg>
</div>
`;

exports[`spinner should render a spinner 1`] = `
<div>
  <svg
    aria-labelledby="testId"
    role="img"
    viewBox="0 0 58 58"
    xmlns="http://www.w3.org/2000/svg"
  >
    <title
      id="testId"
    >
      Working
    </title>
    <g
      fill="none"
      fill-rule="evenodd"
    >
      <g
        stroke="#FFFFFF"
        stroke-width="1.5"
        transform="translate(2 1)"
      >
        <circle
          cx="42.601"
          cy="11.462"
          fill="#FFFFFF"
          fill-opacity="1"
          r="5"
        >
          <animate
            attributeName="fill-opacity"
            begin="0s"
            calcMode="linear"
            dur="1.3s"
            repeatCount="indefinite"
            values="1;0;0;0;0;0;0;0"
          />
        </circle>
        <circle
          cx="49.063"
          cy="27.063"
          fill="#FFFFFF"
          fill-opacity="0"
          r="5"
        >
          <animate
            attributeName="fill-opacity"
            begin="0s"
            calcMode="linear"
            dur="1.3s"
            repeatCount="indefinite"
            values="0;1;0;0;0;0;0;0"
          />
        </circle>
        <circle
          cx="42.601"
          cy="42.663"
          fill="#FFFFFF"
          fill-opacity="0"
          r="5"
        >
          <animate
            attributeName="fill-opacity"
            begin="0s"
            calcMode="linear"
            dur="1.3s"
            repeatCount="indefinite"
            values="0;0;1;0;0;0;0;0"
          />
        </circle>
        <circle
          cx="27"
          cy="49.125"
          fill="#FFFFFF"
          fill-opacity="0"
          r="5"
        >
          <animate
            attributeName="fill-opacity"
            begin="0s"
            calcMode="linear"
            dur="1.3s"
            repeatCount="indefinite"
            values="0;0;0;1;0;0;0;0"
          />
        </circle>
        <circle
          cx="11.399"
          cy="42.663"
          fill="#FFFFFF"
          fill-opacity="0"
          r="5"
        >
          <animate
            attributeName="fill-opacity"
            begin="0s"
            calcMode="linear"
            dur="1.3s"
            repeatCount="indefinite"
            values="0;0;0;0;1;0;0;0"
          />
        </circle>
        <circle
          cx="4.938"
          cy="27.063"
          fill="#FFFFFF"
          fill-opacity="0"
          r="5"
        >
          <animate
            attributeName="fill-opacity"
            begin="0s"
            calcMode="linear"
            dur="1.3s"
            repeatCount="indefinite"
            values="0;0;0;0;0;1;0;0"
          />
        </circle>
        <circle
          cx="11.399"
          cy="11.462"
          fill="#FFFFFF"
          fill-opacity="0"
          r="5"
        >
          <animate
            attributeName="fill-opacity"
            begin="0s"
            calcMode="linear"
            dur="1.3s"
            repeatCount="indefinite"
            values="0;0;0;0;0;0;1;0"
          />
        </circle>
        <circle
          cx="27"
          cy="5"
          fill="#FFFFFF"
          fill-opacity="0"
          r="5"
        >
          <animate
            attributeName="fill-opacity"
            begin="0s"
            calcMode="linear"
            dur="1.3s"
            repeatCount="indefinite"
            values="0;0;0;0;0;0;0;1"
          />
        </circle>
      </g>
    </g>
  </svg>
</div>
`;
